app/models/user.rb

```diff
@@ -0,0 +1,18 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+
+  has_secure_password
+
+  validates :email, presence: true, length: { maximum: Settings.max_email },
+                    format: { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  validates :name, presence: true, length: { maximum: Settings.max_name }
+  validates :password, presence: true, length: { minimum: Settings.min_pass }
+
+  before_save :email_downcase
+
+  private
+   def email_downcase
+    self.email = email.downcase
```

Comment:
```suggestion
    email.downcase!
```
so sánh giúp a downcase! vs downcase nhé

Reply:
downcase!: thay đổi chuỗi thành chữ thường trực tiếp không cần phải gán.
downcase:thay đổi chuỗi thành chữ thường ngay lúc thực thi,nên cần phải gán để lưu chuỗi.

Reply:
còn gì nữa k e ?

Reply:
downcase!:sẽ thay đổi biến gốc,ngược lại downcase thì không

Reply:
thêm nữa là downcase! sẽ trả về nil nếu k tìm dc ký tự hoa nào nhé

Reply:
dạ